core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java

```diff
@@ -0,0 +1,134 @@
+package org.infinispan.conflict.resolution;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.infinispan.Cache;
+import org.infinispan.commands.CommandsFactory;
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.container.DataContainer;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.statetransfer.InboundTransferTask;
+import org.infinispan.statetransfer.StateChunk;
+import org.infinispan.topology.CacheTopology;
+import org.infinispan.topology.LocalTopologyManager;
+
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * @author Ryan Emerson
+ * @since 9.0
+ */
+public class StateReceiverImpl<K, V> implements StateReceiver<K, V> {
+
+   private static Log log = LogFactory.getLog(StateReceiverImpl.class);
+
+   private Cache cache;
+   private CommandsFactory commandsFactory;
+   private Configuration configuration;
+   private DataContainer<K, V> dataContainer;
+   private LocalTopologyManager localTopologyManager;
+   private RpcManager rpcManager;
+   private long transferTimeout;
+
+   @GuardedBy("keyReplicaMapLock")
+   private final Map<K, Map<Address, CacheEntry<K, V>>> keyReplicaMap = new HashMap<>();
+
+   @GuardedBy("keyReplicaMapLock")
+   private final Map<Address, InboundTransferTask> transferTaskMap = new HashMap<>();
+
+   private volatile CompletableFuture<Void> transferFuture = new CompletableFuture<>();
+
+   @Inject
+   public void init(Cache cache,
+                    CommandsFactory commandsFactory,
+                    Configuration configuration,
+                    DataContainer<K, V> dataContainer,
+                    LocalTopologyManager localTopologyManager,
+                    RpcManager rpcManager) {
+      this.cache = cache;
+      this.commandsFactory = commandsFactory;
+      this.configuration = configuration;
+      this.dataContainer = dataContainer;
+      this.localTopologyManager = localTopologyManager;
+      this.rpcManager = rpcManager;
+
+      this.transferTimeout = configuration.clustering().stateTransfer().timeout();
+   }
+
+   @Override
+   public CompletableFuture<List<Map<Address, CacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId) {
```

Comment:
@rvansa I agree. How do you think we should handle such a case? As statereceiver is created on a per cache basis and is only invoked via DefaultConflictResolutionManager::getConflicts(), I think it might be overkill to allow for concurrent requests on this method. Furthermore, I'm not sure allowing concurrent executions of `getConflicts` on the same node should be allowed as this could cause problems when it comes to conflict resolution.

Is it sufficient to simply throw an IllegalStateException in the event that concurrent executions of getConflict (and also getAllReplicasForSegment) occur?